File: 2020/day4/day4.js
const fs = require('fs');
const path = require('path');

function parsePassports(lines) {
  const passports = [{}];
  for (const line of lines) {
    if (line.length === 0) {
      passports.push({});
      continue;
    }
    const entries = line.split(' ');
    const lastPassport = passports[passports.length - 1];
    for (const entry of entries) {
      const [key, value] = entry.split(':')
      lastPassport[key] = value;
    }
  }
  return passports;
}

function part1(input) {
  const lines = input
    .trim()
    .split('\n');
  const requiredFields = [
    'byr',
    'iyr',
    'eyr',
    'hgt',
    'hcl',
    'ecl',
    'pid',
  ];
  const passports = parsePassports(lines);

  const validPassports = passports.filter(p => {
    const keys = Object.keys(p);
    for (const field of requiredFields) {
      if (!keys.includes(field)) return false;
    }
    return true;
  })
  return validPassports.length;
}

function part2(input) {
  const lines = input
    .trim()
    .split('\n');
  const requiredFields = {
    'byr': (value) => parseInt(value) >= 1920 && parseInt(value) <= 2002,
    'iyr': (value) => parseInt(value) >= 2010 && parseInt(value) <= 2020,
    'eyr': (value) => parseInt(value) >= 2020 && parseInt(value) <= 2030,
    'hgt': (value) => {
      const [, num, unit] = value.match(/^(\d+)(\w+)$/) ?? [];
      if (unit === 'cm') return parseInt(num) >= 150 && parseInt(num) <= 193;
      if (unit === 'in') return parseInt(num) >= 59 && parseInt(num) <= 76;
      return false;
    },
    'hcl': (value) => !!value.match(/^\#([0-9a-f]{6})$/),
    'ecl': (value) => ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'].includes(value),
    'pid': (value) => !!value.match(/^(\d{9})$/),
  };
  const passports = parsePassports(lines);

  const validPassports = passports.filter(p => {
    const keys = Object.keys(p);
    const requiredKeys = Object.keys(requiredFields)
    for (const field of requiredKeys) {
      if (!keys.includes(field)) return false;
      if (!requiredFields[field](p[field])) return false;
    }
    return true;
  })
  return validPassports.length;
}

const input = fs.readFileSync(path.resolve(__dirname, './input.txt'), 'utf8');
console.log('day4 part1:', part1(input));
console.log('day4 part2:', part2(input));

/*
const input = require('../input');
const log = console.log;
const compose = (...fns) => (args) => fns.reduceRight((arg, fn) => fn(arg), args);
const split = (on) => (value) => value.split(on);

const splitOnBlankLine = split('\n\n');
const splitOnFieldSeparator = split(':');
const splitOnNewLine = split(/\s+/);

const parseToRecord = (passport) =>
  passport.reduce((record, field) => {
    const [key, value] = splitOnFieldSeparator(field);
    return { ...record, [key]: value };
  }, {});

const isWithinRange = (min, max) => (number) => number >= min && number <= max;
const isValidHeight = ([_, height, unit]) => (unit === 'cm' ? isWithinRange(150, 193) : isWithinRange(59, 76))(height);
const match = (pattern) => (string) => string.match(pattern) ?? [];
const matches = (pattern) => (string) => pattern.test(string);
const isAlwaysValid = () => true;

const validator = {
  byr: isWithinRange(1920, 2002),
  iyr: isWithinRange(2010, 2020),
  eyr: isWithinRange(2020, 2030),
  hgt: compose(isValidHeight, match(/^(\d+)(cm|in)$/)),
  hcl: matches(/^#[0-9a-f]{6}$/),
  ecl: matches(/^(amb|blu|brn|gry|grn|hzl|oth)$/),
  pid: matches(/^\d{9}$/),
  cid: isAlwaysValid,
};

const hasAllRequiredFields = (passport) =>
  ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'].every((field) => passport[field]);
const hasAllValidFields = (passport) => Object.entries(passport).every(([key, value]) => validator[key](value));

const passports = splitOnBlankLine(input(__dirname, './input.txt')).map(splitOnNewLine).map(parseToRecord);

const solutionOne = passports.filter(hasAllRequiredFields).length;
log(`Solution pt.1 ${solutionOne}`);

const solutionTwo = passports.filter(hasAllRequiredFields).filter(hasAllValidFields).length;
log(`Solution pt.2 ${solutionTwo}`);
*/

File: 2020/day4/day4.clj
'{:dependencies [[org.clojure/clojure "1.10.1"]]}
(require '[clojure.string :as str])

(def input
  (->>
   (slurp "2020/day4/input.txt")
   (#(str/split % #"\n\n"))))

(defn parsePassports
  "Take input and parse into vector of passport map"
  [input]
  (->>
   input
   (map #(str/split %1 #"\n| "))
   (map
    (fn [entry] (reduce #(into %1 [(str/split %2 #":")]) {} entry)))))

(def passports (parsePassports input))

;; (defn includes? [list val] (some? (some (partial = val) list)))
;; (defn includes? [list val] (some #(= val %1) list))
(defn includes? [list val] (contains? (set list) val))

;; Use (Integer. val) to make sure value is int
(defn between? "Check if val is between low and high" [low high val] (<= low (Integer/parseInt (str val)) high))

(def req-fields ["byr" "iyr" "eyr" "hgt" "hcl" "ecl" "pid"])

(def validator {"byr" (partial between? 1920 2002)
                "iyr" (partial between? 2010 2020)
                "eyr" (partial between? 2020 2030)
                "hgt" (fn [val] (let [[_ num unit] (re-matches #"^(\d+)(\w+)$" val)]
                                  (cond
                                    (= unit "cm") (between? 150 193 num)
                                    (= unit "in") (between? 59 76 num)
                                    :else false)))
                "hcl" (comp some? (partial re-matches #"^\#([0-9a-f]{6})$"))
                "ecl" (partial includes? ["amb" "blu" "brn" "gry" "grn" "hzl" "oth"])
                "pid" (comp some? (partial re-matches #"^(\d{9})$"))
                "cid" (fn [_] true)})

(defn has-req-fields? [passport] (every? (set (keys passport)) req-fields))

(defn is-valid? [passport] (every? #((get validator (first %1)) (second %1)) (seq passport)))

(defn part1 []
  (prn "day4 part1:"
       (->>
        passports
        (filter has-req-fields?)
        (count))))

(defn part2 []
  (prn "day4 part2:"
       (->>
        passports
        (filter #(and (has-req-fields? %) (is-valid? %)))
        (count))))

(part1)
(part2)

;; (defn parsePassports
;;   "Take input and parse into vector of passport map"
;;   [input]
;;   (->>
;;    input
;;    (map (fn [entry]
;;           (into {} (map (comp vec next) (re-seq #"(\w{3}):(\S+)" entry)))))
;;    vec))

;; --------------------------------------------------------------------------------

;; (def transpose
;;   (partial apply map vector))

;; (defn parse [passport]
;;   (let [[ks vs] (->> passport
;;                      (re-seq #"([^ \n]+):([^ \n]+)")
;;                      (map rest)
;;                      transpose)]
;;     (zipmap ks vs)))

;; (def passports
;;   (->> #"\n\n"
;;        (string/split (slurp "resources/2020/day4.txt"))
;;        (map parse)))

;; --------------------------------------------------------------------------------

;; (->>
;;      (string/split (get-res input-file) #"\n\n")

;;      ;; form:
;;      (map #(string/replace % #"\n" " "))
;;      (map #(string/split % #" "))

;;      ;; now, we have a collection of strings "key:val", so turn them into proper maps (still string-> string):
;;      (map
;;       (fn [coll]
;;         (reduce
;;          #(let [[key val] (string/split %2 #":")]
;;             (assoc %1 key val))
;;          {} coll))))

;; --------------------------------------------------------------------------------

;; (def mandatory-keys ["byr" "iyr" "eyr" "hgt" "hcl" "ecl" "pid"])

;; (defn parse-input [str]
;;   (->> (str/split str #"\n\n")
;;        (map (fn [str]
;;               (->> (str/split str #"[\n| ]")
;;                    (map #(str/split % #":"))
;;                    (reduce #(assoc %1 (first %2) (second %2)) {}))))))

;; (defn valid-keys? [passport]
;;   (every? passport mandatory-keys))

;; (defn part-1 [file-name]
;;   (let [passports (parse-input (slurp file-name))]
;;     (count (filter valid-keys? passports))))

;; (defn parse-height [height-str suffix]
;;   (Integer/parseInt
;;    (.substring height-str 0 (str/index-of height-str suffix))))

;; (defn valid-height? [height-str]
;;   (cond
;;     (str/ends-with? height-str "cm") (<= 150 (parse-height height-str "cm") 193)
;;     (str/ends-with? height-str "in") (<= 59 (parse-height height-str "in") 76)
;;     :else false))

;; (defn valid-color? [color-str]
;;   (if-not (str/starts-with? color-str "#")
;;     false
;;     (let [color-values (.substring color-str 1)]
;;       (every? (fn [char-val]
;;                 (or (Character/isDigit char-val)
;;                     (<= (int \a) (int char-val) (int \f))))
;;               color-values))))

;; (def valid-eye-color-values #{"amb" "blu" "brn" "gry" "grn" "hzl" "oth"})

;; (defn valid-eye-color? [eye-color]
;;   (contains? valid-eye-color-values eye-color))

;; (defn valid-pid? [str-pid]
;;   (and (= 9 (count str-pid))
;;        (every? (fn [char]
;;                  (Character/isDigit char)) str-pid)))

;; (defn valid-value? [[key val]]
;;   (case key
;;     "byr" (<= 1920 (Integer/parseInt val) 2002)
;;     "iyr" (<= 2010 (Integer/parseInt val) 2020)
;;     "eyr" (<= 2020 (Integer/parseInt val) 2030)
;;     "hgt" (valid-height? val)
;;     "hcl" (valid-color? val)
;;     "ecl" (valid-eye-color? val)
;;     "pid" (valid-pid? val)
;;     "cid" true))

;; (defn valid-values? [passport]
;;   (every? valid-value? passport))

;; (defn part-2 [file-name]
;;   (let [passports (parse-input (slurp file-name))]
;;     (count (filter (fn [passport]
;;                      (and (valid-keys? passport)
;;                           (valid-values? passport)))
;;                    passports))))
